UnityProject/Assets/Scripts/Core/ConfigurationSaves/ConfigurationSaves.cs

```diff
@@ -0,0 +1,302 @@
+				if (string.IsNullOrEmpty(CashedForkName))
+				{
+					var path = Path.GetFullPath(Path.Combine(Application.streamingAssetsPath,
+						"Config",
+						"buildinfo.json"));
+					var text = File.ReadAllText(path);
+					var data = JsonConvert.DeserializeObject<BuiltFork>(text);
+					if (data == null)
+					{
+						CashedForkName = "Unitystation";
+					}
+					else
+					{
+						CashedForkName = data.ForkName;
+					}
+				}
+
+				return CashedForkName;
+			}
+		}
+
+		private static Dictionary<string, FileSystemWatcher> CurrentlyWatchingFile = new Dictionary<string, FileSystemWatcher>();
+
+		private static Dictionary<Action, string> RegisteredToFile = new Dictionary<Action, string>();
+
+		private static Dictionary<string, List<Action>> RegisteredToWatch = new Dictionary<string, List<Action>>();
+
+		private class BuiltFork
+		{
+			public string ForkName; // = Unitystation"
+		}
+
+
+		private static readonly string[] AllowedExtensions = new[] {".txt", ".json", ".toml", ".yaml", ".data", ".log"};
+
+		private static string ValidatePath(string relativePath, AccessCategory accessCategory, bool userPersistent,
+			bool CreateFile, bool AddExtension = true)
+		{
+			bool isAllowedExtension = false;
+			var extension = "";
+
+			if (AddExtension)
+			{
+				foreach (var allowedExtension in AllowedExtensions)
+				{
+					if (relativePath.EndsWith(allowedExtension))
+					{
+						isAllowedExtension = true;
+						break;
+					}
+				}
+
+				if (isAllowedExtension == false)
+				{
+					switch (accessCategory)
+					{
+						case AccessCategory.Config:
+							extension = ".txt";
+							break;
+						case AccessCategory.Data:
+							extension = ".Data";
+							break;
+						case AccessCategory.Logs:
+							extension = ".log";
+							break;
+					}
+				}
+
+			}
+
+			string resolvedPath = "";
+
+			if (userPersistent)
+			{
+				resolvedPath = Path.GetFullPath(Path.Combine(Application.persistentDataPath, ForkName , accessCategory.ToString(), relativePath + extension));
+				if (resolvedPath.StartsWith(Path.GetFullPath(Path.Combine(Application.persistentDataPath, ForkName, accessCategory.ToString()))) == false)
+				{
+					Logger.LogError($"Malicious PATH was passed into File access, HEY NO! Stop being naughty with the PATH! {resolvedPath}");
+					throw new Exception($"Malicious PATH was passed into File access, HEY NO! Stop being naughty with the PATH! {resolvedPath}");
+				}
+			}
+			else
+			{
+				resolvedPath = Path.GetFullPath(Path.Combine(Application.streamingAssetsPath,accessCategory.ToString(), relativePath + extension));
+				if (resolvedPath.StartsWith(Path.GetFullPath(Path.Combine(Application.streamingAssetsPath,  accessCategory.ToString()))) == false)
+				{
+					Logger.LogError($"Malicious PATH was passed into File access, HEY NO! Stop being naughty with the PATH! {resolvedPath}");
+					throw new Exception($"Malicious PATH was passed into File access, HEY NO! Stop being naughty with the PATH! {resolvedPath}");
```

Comment:
It might be good to give these two different error messages, would make it easier to see where the problem is when looking at a log.

Reply:
what will be best to do putting exception? at the beginning